src/sdc11073/wsdiscovery/wsdimpl.py

```diff
@@ -0,0 +1,620 @@
+    from sdc11073.xml_types.msg_types import MessageType
+    from sdc11073.location import SdcLocation
+    from collections.abc import Iterable
+
+
+WSA_ANONYMOUS = nsh.WSA.namespace + '/anonymous'
+ADDRESS_ALL = "urn:docs-oasis-open-org:ws-dd:ns:discovery:2009:01"  # format acc to RFC 2141
+
+NS_D = nsh.WSD.namespace
+MATCH_BY_LDAP = NS_D + '/ldap'  # "http://docs.oasis-open.org/ws-dd/ns/discovery/2009/01/ldap"
+MATCH_BY_URI = NS_D + '/rfc3986'  # "http://docs.oasis-open.org/ws-dd/ns/discovery/2009/01/rfc3986"
+MATCH_BY_UUID = NS_D + '/uuid'  # "http://docs.oasis-open.org/ws-dd/ns/discovery/2009/01/uuid"
+MATCH_BY_STRCMP = NS_D + '/strcmp0'  # "http://docs.oasis-open.org/ws-dd/ns/discovery/2009/01/strcmp0"
+
+
+def types_info(types: list[QName]) -> list[str]:
+    """Make printable strings from list of QNames (helper method for logging)."""
+    return [str(t) for t in types] if types else types
+
+
+def match_scope(my_scope: str, other_scope: str, match_by: str) -> bool:
+    """match_scope checks if my_scope matches other_scope by applying the algorithm defined by match_by.
+
+    match_scope correctly handles "%2F" (== '/') encoded values.
+    """
+    if match_by in (MATCH_BY_LDAP, MATCH_BY_URI, MATCH_BY_UUID, '', None):
+        my_scope = urlsplit(my_scope)
+        other_scope = urlsplit(other_scope)
+        if my_scope.scheme.lower() != other_scope.scheme.lower() \
+                or my_scope.netloc.lower() != other_scope.netloc.lower():
+            return False
+        if my_scope.path == other_scope.path:
+            return True
+        src_path_elements = my_scope.path.split('/')
+        target_path_elements = other_scope.path.split('/')
+        src_path_elements = [unquote(elem) for elem in src_path_elements]
+        target_path_elements = [unquote(elem) for elem in target_path_elements]
+        if len(src_path_elements) > len(target_path_elements):
+            return False
+        return all(target_path_elements[i] == elem for i, elem in enumerate(src_path_elements))
+    if match_by == MATCH_BY_STRCMP:
+        return my_scope == other_scope
+    return False
+
+
+def match_type(type1: QName, type2: QName) -> bool:
+    """Check if namespace and localname are identical."""
+    return type1.namespace == type2.namespace and type1.localname == type2.localname
+
+
+def _is_type_in_list(ttype: QName, types: list[QName]) -> bool:
+    return any(match_type(ttype, entry) for entry in types)
+
+
+def _is_scope_in_list(uri: str, match_by: str, srv_sc: wsd_types.ScopesType) -> bool:
+    # returns True if every entry in scope.text is also found in srv_sc.text
+    # all entries are URIs
+    if srv_sc is None:
+        return False
+    return any(match_scope(uri, entry, match_by) for entry in srv_sc.text)
+
+
+def matches_filter(service: Service,
+                   types: list[QName] | None,
+                   scopes: wsd_types.ScopesType | None,
+                   logger: Logger | None = None) -> bool:

```

Comment:
remove `logger` argument. I dont think its the job of `matches_filter` to log it